openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,632 @@
+    then take it as the 100%, which mean an account.move.line both specify '2' as the
+    analytic.account.id and it has 1 analytic.tag also have that analytic.account then
+    the value will sum together
+    """
+    if not openupgrade.column_exists(env.cr, table_name, "analytic_distribution"):
+        openupgrade.logged_query(
+            env.cr,
+            f"""
+            ALTER TABLE {table_name}
+            ADD COLUMN IF NOT EXISTS analytic_distribution jsonb;
+            """,
+        )
+    is_having_analytic_account_id_col = openupgrade.column_exists(
+        env.cr, table_name, "analytic_account_id"
+    )
+    # Build query
+    select_query = f"""
+        SELECT
+        {table_name}.id,
+        unnest(array_agg(tmp.tag_id)) AS tag,
+        unnest(array_agg(tmp.account_id)) AS analytic_account_of_tag,
+        unnest(array_agg(tmp.percentage)) AS percentage
+    """
+    groupby_query = f"""
+        GROUP BY
+            {table_name}.id
+    """
+    if is_having_analytic_account_id_col:
+        select_query += f"""
+            , {table_name}.analytic_account_id AS account_analytic_account
+        """
+        groupby_query += f"""
+            , {table_name}.analytic_account_id
+        """
+    openupgrade.logged_query(
+        env.cr,
+        f"""
+        WITH sub AS (
+            WITH tmp AS (
+                SELECT
+                    id,
+                    account_id,
+                    percentage,
+                    tag_id
+                FROM
+                account_analytic_distribution
+            )
+            {select_query}
+            FROM
+                {table_name}
+                JOIN account_analytic_tag_{table_name}_rel rel
+                    ON {table_name}.id = rel.{table_name}_id
+                JOIN account_analytic_tag aat
+                    ON aat.id = rel.account_analytic_tag_id
+                JOIN tmp
+                    ON tmp.tag_id = aat.id
+            {groupby_query}
+        )
+        UPDATE {table_name}
+        SET analytic_distribution = analytic_distribution_sub.analytic_distribution
+        FROM (
+            SELECT
+                id,
+                jsonb_object_agg(analytic_account_of_tag::text, total_percentage)
+                AS analytic_distribution
+            FROM (
+                SELECT
+                    sub.id,
+                    analytic_account_of_tag,
+                    sum(percentage) AS total_percentage
+                FROM
+                    sub
+                GROUP BY
+                    sub.id,
+                    analytic_account_of_tag
+            ) AS aggregated_data
+            GROUP BY id
+        ) AS analytic_distribution_sub
+        WHERE {table_name}.id = analytic_distribution_sub.id
+        """,
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    # Perform rename stuff
+    openupgrade.rename_xmlids(env, _xmlids_renames)
```

Comment:
```suggestion
    openupgrade.rename_xmlids(env.cr, _xmlids_renames)
```

Otherwise, I am getting this error:

```
site-packages/openupgradelib/openupgrade.py", line 1236, in rename_xmlids
    cr.execute(get_data_query, tuple(old.split(".")))
    ^^^^^^^^^^
AttributeError: 'Environment' object has no attribute 'execute'
```

https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L1211

The code of the latest version of openupgradelib is like this:

![image](https://github.com/OCA/OpenUpgrade/assets/1765420/879bc25c-b66f-4299-bb4c-0224024ae998)

Reply:
Sorry about that, i have fixed